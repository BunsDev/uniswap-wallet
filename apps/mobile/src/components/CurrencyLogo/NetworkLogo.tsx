import { SpacingProps, SpacingShorthandProps } from '@shopify/restyle'
import React from 'react'
import { Image, StyleSheet } from 'react-native'
import { style } from 'src/components/CurrencyLogo/styles'
import { Box } from 'src/components/layout/Box'
import { iconSizes } from 'ui/src/theme/iconSizes'
import { Theme } from 'ui/src/theme/restyle/theme'
import { ChainId, CHAIN_INFO } from 'wallet/src/constants/chains'

type NetworkLogoProps = {
  chainId: ChainId
  size?: number
  shape?: 'circle' | 'square'
} & SpacingProps<Theme> &
  SpacingShorthandProps<Theme>

export const SQUARE_BORDER_RADIUS = 6

export function TransactionSummaryNetworkLogo({
  chainId,
  size = iconSizes.icon20,
}: Pick<NetworkLogoProps, 'chainId' | 'size'>): JSX.Element {
  return (
    <Box borderColor="background0" style={styles.squareLogoOutline}>
      <NetworkLogo chainId={chainId} shape="square" size={size} />
    </Box>
  )
}

function _NetworkLogo({
  chainId,
  shape = 'circle',
  size = iconSizes.icon20,
}: NetworkLogoProps): JSX.Element | null {
  const { logo } = CHAIN_INFO[chainId]
  const borderRadius = shape === 'circle' ? size / 2 : SQUARE_BORDER_RADIUS
  return logo ? (
    <Box style={[{ borderRadius }, styles.iconWrapper]}>
      <Image source={logo} style={[style.image, { width: size, height: size }]} />
    </Box>
  ) : null
}

export const NetworkLogo = React.memo(_NetworkLogo)

const styles = StyleSheet.create({
  iconWrapper: {
    overflow: 'hidden',
  },
  squareLogoOutline: {
    borderRadius: 7, // when inner icon borderRadius = 6 and size = 20
    borderWidth: 2,
  },
})
