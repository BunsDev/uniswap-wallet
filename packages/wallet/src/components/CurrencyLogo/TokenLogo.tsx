import { Image } from 'react-native'
import { useTheme } from 'tamagui'
import { Box, Text } from 'ui/src'
import { iconSizes } from 'ui/src/theme'
import { isSVGUri, uriToHttp } from 'utilities/src/format/urls'
import { STATUS_RATIO } from 'wallet/src/components/CurrencyLogo/CurrencyLogo'
import { SHADOW_OFFSET, style, THIN_BORDER } from 'wallet/src/components/CurrencyLogo/styles'
import { ChainId } from 'wallet/src/constants/chains'
import { RemoteSvg } from 'wallet/src/features/images/RemoteSvg'
import { NetworkLogo } from './NetworkLogo'

interface TokenLogoProps {
  url?: string | null
  symbol?: string
  chainId?: ChainId
  size?: number
  hideNetworkLogo?: boolean
}

export function TokenLogo({
  url,
  symbol,
  chainId,
  size = iconSizes.icon40,
  hideNetworkLogo,
}: TokenLogoProps): JSX.Element {
  const theme = useTheme()
  const showNetworkLogo = !hideNetworkLogo && chainId && chainId !== ChainId.Mainnet
  const httpUri = url ? uriToHttp(url)[0] : null

  let tokenImage = null

  if (httpUri) {
    if (isSVGUri(httpUri)) {
      tokenImage = (
        <Box borderRadius={size / 2} overflow="hidden">
          <RemoteSvg
            backgroundColor={theme.surface3.get()}
            borderRadius={size / 2}
            height={size}
            imageHttpUrl={httpUri}
            width={size}
          />
        </Box>
      )
    } else {
      tokenImage = (
        <Image
          source={{ uri: httpUri }}
          style={[
            style.image,
            {
              backgroundColor: theme.surface3.get(),
              width: size,
              height: size,
              borderRadius: size / 2,
              borderColor: theme.surface3.get(),
              borderWidth: THIN_BORDER,
            },
          ]}
        />
      )
    }
  }

  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      {httpUri ? (
        tokenImage
      ) : (
        <Box
          alignItems="center"
          bg="$surface3"
          borderRadius="$roundedFull"
          height={size}
          justifyContent="center"
          px="$spacing8"
          width={size}>
          <Text
            adjustsFontSizeToFit
            color="$neutral1"
            flex={0}
            numberOfLines={1}
            textAlign="center">
            {symbol?.slice(0, 3)}
          </Text>
        </Box>
      )}
      {showNetworkLogo && (
        <Box
          bottom={0}
          position="absolute"
          right={0}
          shadowColor="black"
          shadowOffset={SHADOW_OFFSET}
          shadowOpacity={0.1}
          shadowRadius={2}>
          <NetworkLogo chainId={chainId} size={size * STATUS_RATIO} />
        </Box>
      )}
    </Box>
  )
}
