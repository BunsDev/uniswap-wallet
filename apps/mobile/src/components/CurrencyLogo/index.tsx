import React from 'react'
import { TokenLogo } from 'src/components/CurrencyLogo/TokenLogo'
import { iconSizes } from 'ui/src/theme/iconSizes'
import { CurrencyInfo } from 'wallet/src/features/dataApi/types'

interface CurrencyLogoProps {
  currencyInfo: Maybe<CurrencyInfo>
  size?: number
  hideNetworkLogo?: boolean
}

export const STATUS_RATIO = 2 / 5

export function CurrencyLogo({
  currencyInfo,
  size = iconSizes.icon40,
  hideNetworkLogo,
}: CurrencyLogoProps): JSX.Element | null {
  if (!currencyInfo) return null

  const { currency, logoUrl } = currencyInfo
  const { chainId, symbol } = currency

  return (
    <TokenLogo
      chainId={chainId}
      hideNetworkLogo={hideNetworkLogo}
      size={size}
      symbol={symbol}
      url={logoUrl}
    />
  )
}
