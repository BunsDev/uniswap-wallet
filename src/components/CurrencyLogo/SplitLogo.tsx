import React from 'react'
import { useAppTheme } from 'src/app/hooks'
import { CurrencyLogo, STATUS_RATIO } from 'src/components/CurrencyLogo'
import { NetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { Box } from 'src/components/layout/Box'
import { ChainId } from 'src/constants/chains'
import { CurrencyInfo } from 'src/features/dataApi/types'
import { getNetworkForegroundColor } from 'src/utils/colors'

interface Props {
  inputCurrencyInfo: NullUndefined<CurrencyInfo>
  outputCurrencyInfo: NullUndefined<CurrencyInfo>
  size: number
  chainId: ChainId | null
}

/*
 * Logo, where left 50% of width is taken from one icon (its left 50%)
 * and right side is taken from another icon (its right 50%)
 */
export function SplitLogo({
  size,
  inputCurrencyInfo,
  outputCurrencyInfo,
  chainId,
}: Props): JSX.Element {
  const theme = useAppTheme()

  const iconSize = size / 2

  const icon =
    chainId && chainId !== ChainId.Mainnet ? (
      <NetworkLogo
        backgroundColor={getNetworkForegroundColor(theme, chainId)}
        borderColor={theme.colors.background0}
        borderRadius="rounded8"
        borderWidth={2}
        chainId={chainId}
        size={size * STATUS_RATIO}
      />
    ) : undefined

  return (
    <Box height={size} width={size}>
      <Box
        left={0}
        overflow="hidden"
        position="absolute"
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={inputCurrencyInfo} size={size} />
      </Box>
      <Box
        flexDirection="row-reverse"
        overflow="hidden"
        position="absolute"
        right={0}
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={outputCurrencyInfo} size={size} />
      </Box>
      {icon && (
        <Box bottom={-4} position="absolute" right={-4}>
          {icon}
        </Box>
      )}
    </Box>
  )
}
