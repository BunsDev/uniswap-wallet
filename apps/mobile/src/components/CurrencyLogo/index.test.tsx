import React from 'react'
import { render } from 'src/test/test-utils'
import { CurrencyLogo } from 'wallet/src/components/CurrencyLogo/CurrencyLogo'
import { arbitrumDaiCurrencyInfo, uniCurrencyInfo } from 'wallet/src/test/fixtures'

jest.mock('ui/src/assets/', () => 'ethereum-logo.png')

it('renders a currency logo without network logo', () => {
  const tree = render(<CurrencyLogo currencyInfo={uniCurrencyInfo} size={20} />)
  expect(tree).toMatchSnapshot()
})

it('renders a currency logo with network logo', () => {
  const tree = render(<CurrencyLogo currencyInfo={arbitrumDaiCurrencyInfo} size={20} />)
  expect(tree).toMatchSnapshot()
})
