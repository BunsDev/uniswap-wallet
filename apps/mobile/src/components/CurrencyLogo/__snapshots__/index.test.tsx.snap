// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a currency logo with network logo 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "height": 20,
      "justifyContent": "center",
      "width": 20,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      [
        {
          "resizeMode": "contain",
        },
        {
          "backgroundColor": "#2222220D",
          "borderColor": "#2222220D",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    style={
      {
        "alignItems": "stretch",
        "bottom": 0,
        "flexDirection": "column",
        "position": "absolute",
        "right": 0,
        "shadowColor": "rgb(0,0,0)",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 2,
      }
    }
  />
</View>
`;

exports[`renders a currency logo without network logo 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "height": 20,
      "justifyContent": "center",
      "width": 20,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      [
        {
          "resizeMode": "contain",
        },
        {
          "backgroundColor": "#2222220D",
          "borderColor": "#2222220D",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
</View>
`;
